A34-weighted-round-robin.md

```diff
@@ -0,0 +1,111 @@
+
+## Proposal
+
+The proposal is to carry [`load_balancing_weight`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L108) of [`LbEndpoint`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L76) from [`ClusterLoadAssignment`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint.proto#L34) response to `lb_policy` and implement a `weighted_round_robin` policy based on Earliest deadline first scheduling algorithm picker [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling). Each endpoint will get fraction equal to the [`load_balancing_weight`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L108) for the [`LbEndpoint`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L76) divided by the sum of the `load_balancing_weight` of all `LbEndpoint` within the same [`LocalityLbEndpoints`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L116) of traffic routed to the locality.
+
+### Overview of `weighted_round_robin` policy
+
+`weighted_round_robin` policy is powered by [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) picker. EDF picker maintains a priority queue of `EdfEntry`. On the top of the queue, it’s the entry with lowest `deadline`. `order_offset` is the tie breaker when two entries have same deadline to maintain FIFO order.
+
+Proposed `EdfEntry`
+```
+struct EdfEntry {
+    // primary key for the priority queue. The entry with least deadline is the top of the queue.
+    double deadline;
+    // secondary key for the priority queue. Used as a tiebreaker for same deadline to maintain FIFO order.
+    uint64 order_offset;
+    // `load_balancing_weight` of this endpoint from the latest `ClusterLoadAssignment`.
+    double weight;
+    // Subchannel data structure of this endpoint.
+    Subchannel subchannel;
+}
+```
+
+- On each call to the `Pick`, EDF picker picks the entry `e` on the top of the queue, returns the subchannel associated with the entry. After that, picker updates the `deadline` of `e` to `e.deadline + 1/weight` and either performs a pop and push the entry back to the queue or key increase operation.
+- `weighted_round_robin` updates the entries in EDF priority queue on any change of subchannel `ConnnectiveState` or endpoint `load_balancing_weight`.
+- If all endpoints have the same `load_balancing_weight`, `EDF` picker degenerates to `round_robin` picker. It's easier to reason and consistent with envoy.
+- Endpoints do not have `load_balancing_weight` assigned are discarded.
```

Comment:
I don't think this is the behavior we want.  The comment on the `load_balancing_weight` field says "If unspecified, each host is presumed to have equal weight in a locality."

Reply:
That doesn't really make sense either.  If only one address has a missing weight, what do we assign it?  The average of the defined weights?

Reply:
Good point.  I suggest checking the Envoy implementation to see how they handle this.  I suspect they just assume a weight of 1 if the field is not set.  We can also ask them to improve the comment.

Reply:
https://github.com/envoyproxy/envoy/blob/5d95032baa803f853e9120048b56c8be3dab4b0d/source/common/upstream/upstream_impl.cc#L1243 and https://github.com/envoyproxy/envoy/blob/5d95032baa803f853e9120048b56c8be3dab4b0d/source/common/upstream/upstream_impl.cc#L359

I think you are right. it uses default value which is 0, and in the constructor it's adjusted to 1. I am not sure it's good default behavior because the endpoints with unspecified weight receive same amount of traffic with endpoints specified with 1. However, I don't know how to assign it properly either and I think it's control plane's job to assign proper weight for each endpoint. I will make the proposal same as envoy behavior if nobody has other concerns